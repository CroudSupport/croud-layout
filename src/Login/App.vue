<template>
    <login v-if="active" :view.sync="view" :username.sync="username" transition="bouncer" transition-mode="out-in" @login="$emit('login')"></login>
</template>
<script>
import Login from './Components/Login'

export default {
    components: {
        Login,
    },

    data() {
        return {
            active: false,
            view: 'login',
            username: '',
        }
    },

    mounted() {
        setTimeout(() => {
            this.active = true
        }, 250)
    },
}
</script>
