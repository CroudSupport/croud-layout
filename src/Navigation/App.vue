<template>
  <div class="contained scrollable navigation-block">
    <div class="footer">
      <strong class="expand-button" @click.prevent="toggle"><i :class="toggleClass"></i></strong>
    </div>
    <ul data-depth="1">
      <component
      v-for="item in items"
      :is="item.menuType"
      :user="user"
      :items="items"
      :item="item" depth="1"
      :expanded="expanded"
      />
    </ul>
  </div>
</template>

<script>
    import { mapGetters } from 'vuex'

    import ListItem from './Components/ListItem'
    import ListHeading from './Components/ListHeading'

    export default {
        data() {
            return {}
        },
        components: {
            ListItem, ListHeading,
        },
        props: ['depth', 'expanded'],
        computed: {
            toggleClass() {
                const classNames = ['arrow circle double icon']

                if (this.expanded) {
                    classNames.push('left')
                } else {
                    classNames.push('right')
                }
                return classNames.join(' ')
            },

            items() {
                const stashed = JSON.parse(localStorage.getItem(`main_navigation_${this.user.code}`))

                if (stashed && stashed.data && stashed.data.list) {
                    return stashed.data.list
                }

                return [
                    {
                        label: 'Dashboard',
                        uri: '/',
                        icon: 'dashboard icon',
                        menuType: 'list-item',
                    },
                    {
                        label: 'Croud Academy',
                        uri: '/academy/',
                        icon: 'university icon',
                        menuType: 'list-item',
                        expanded: true,
                        children: [
                            {
                                label: 'My Courses',
                                uri: '/academy/course/owned/',
                                menuType: 'list-item',
                            },
                            {
                                label: 'Available Courses',
                                uri: '/academy/course/available/',
                                menuType: 'list-item',
                            },
                        ],
                    },
                    {
                        label: 'User Management',
                        uri: '/users/manager',
                        icon: 'users icon',
                        menuType: 'list-item',
                        permission: 'menu.user',
                    },
                ]
            },

            ...mapGetters({
                permissions: 'universal/permissions',
                user: 'universal/user',
            }),
        },
        methods: {
            toggle() {
                this.expanded = !this.expanded
            },
            setExpanded(value) {
                document.getElementById('main-content').className = !value ? 'collapsed' : ''
            },
        },

        watch: {
            expanded(value) {
                this.setExpanded(value)
            },
        },
        mounted() {
            this.setExpanded(this.expanded)
            setTimeout(() => {
                $(document.body).removeClass('no-animation')
            }, 0)
        },
    }
</script>
