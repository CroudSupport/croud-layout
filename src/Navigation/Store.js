import storage from 'localstorage'

export default {
    getKey(user) {
        return `main_navigation_${user.id}`
    },
    getNavigation(user, defaultValue = null) {
    // if (storage.exists(this.getKey(user)))
    //   return storage.get(this.getKey(user))
        return defaultValue
    },
    setNavigation(user, data) {
        storage.set(`main_navigation_${user.id}`, data)
    },
}
