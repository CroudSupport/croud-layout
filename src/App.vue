<template>
  <div id="main-container">
    <top-bar v-if="user.id && !suppressTopbar" :user="user"></top-bar>
    <div v-if="loading" class="ui very padded basic center aligned loading-overlay segment">
        <div>
            <img class="ui small image" src='../static/img/croud_logo_new.svg' />
        </div>
    </div>
    <login v-else-if="!user.id"></login>
    <slot v-else name="custom-layout">
        <div id="main-content">
            <div v-if="!suppressNav" id="main-navigation" class="main-navigation">
                <div class="contained">
                    <navigation
                    :expanded="expanded"
                    :items="items"
                    >
                    </navigation>
                </div>
            </div>
            <div :id="!suppressNav ? 'main-content-container' : ''">
                <!-- <div v-if="title" class="headingLinks"><span slot="links"></span></div> -->
                <h1 v-if="title" v-html="title"></h1>
                <div id="main-content-body" v-bind:class="classList">
                    <slot name="content">
                        <h1 class="ui heading">Test</h1>
                    </slot>
                </div>
            </div>
        </div>
    </slot>
  </div>
</template>
<script>
import { mapGetters, mapActions } from 'vuex'

import TopBar from './components/TopBar'
import Navigation from './Navigation/App'
import Login from './Login/App'

export default {
    components: { TopBar, Navigation, Login },
    props: {
        title: {},
        flush: {
            default: true,
        },
        suppressNav: {
            default: false,
        },
        suppressTopbar: {
            default: false,
        },
    },

    computed: {
        classList() {
            const arr = []
            if (this.flush) {
                arr.push('flush-body')
            }
            if (!this.title) {
                arr.push('without-heading')
            }
            return arr.join(' ')
        },

        ...mapGetters({
            user: 'universal/user',
            jwt: 'universal/jwt',
            root: 'universal/root',
            loading: 'universal/loading',
        }),
    },

    methods: {
        ...mapActions({
            updateUser: 'universal/updateUser',
            updatePermissions: 'universal/updatePermissions',
            updateUserSwitches: 'universal/updateUserSwitches',
            authenticate: 'universal/auth',
        }),

        handleLogin() {
            if (!this.jwt.sub) return

            this.authenticate()
        },
    },

    data() {
        return {
            items: [],
            expanded: false,
        }
    },

    mounted() {
        this.handleLogin()
    },
}
</script>

<style lang="scss">
    @import './sass/layout.scss';

    .ui.basic.loading-overlay.segment {
        background-color: rgba(0,0,0,.75);
        position: fixed;
        width: 100vw;
        height: 100vh;
        z-index: 9999;
        margin-top: 0;
        display: flex;
        justify-content: center;
        align-items: center;
        color: white;
    }
</style>
