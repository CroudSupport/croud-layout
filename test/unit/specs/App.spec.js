import Vue from 'vue'
import Vuex from 'vuex'

// import _ from 'lodash'
import localforage from 'localforage'
import localstorage from 'localstorage'
// import axios from 'axios'

import App from '../../../src/App'
import Store from '../../../src/store'

const jwt = 'eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImp0aSI6IjVlNjE0OWQyMjgzYzBhZmEwOWJiZDJiYzE3ZDA2ZDk1ODRjZDcxN2E5M2QwNjhiYTU0NmIyNGI3ODE3MDI2NTY1MjBjYjZiYjFkMTk1MmVlIn0.eyJhdWQiOiJnYXRld2F5LWlyZWxhbmQiLCJqdGkiOiI1ZTYxNDlkMjI4M2MwYWZhMDliYmQyYmMxN2QwNmQ5NTg0Y2Q3MTdhOTNkMDY4YmE1NDZiMjRiNzgxNzAyNjU2NTIwY2I2YmIxZDE5NTJlZSIsImlhdCI6MTUwMDk3NDMwNSwibmJmIjoxNTAwOTc0MzA1LCJleHAiOjE1MDEyMzM1MDUsInN1YiI6IjIxMTciLCJzY29wZXMiOltdfQ.SGcw02bTra8PkAY1k3jbTWWh6kQwxP9CxnOU8wVzyXbVeevSZ9DomHMmMTBKTBajhNOvsrdFqROJj8nAO4NLqoroNAsK8syS1ytKHDmzFtOjDLxx4rEnJgs8jANVPeFAd8EMSD96I6Rqt0v0BmohjJp5H7WXGvvO8MdPsPSVPTW2OVaLjIXM6b8Ra-X0CV-zyeh48ynYQ_dXdIXssZM8s0i-iuDfQPcZwE8j4Yenq-Omm6tkf2ScG5AzS_3M61svsU09wXpBPvsvXsEFOF74Cjy1Sli0hUg90Bz9bX1uW983Dl3VDcsVqBfufR8fqJwjxmSotZUN8pKLUlXoTYD7LHj3UKluZlVDXMHd_J8l659X5M8tH18jIU3E26vz96LaosGJ6yB4aKQsKQAySrEHiB4F_PJOBi09hh1jLvPOi6e_XCdSieml3hZiGid8zEoR77yxxQsBHGY4YAda5Fb93mKy8ckCKbCjXybahrGleq1GI28f3ZwuW_RdhgjMuLIE6zE8mZrbZXNB8dvO7zfgPq1ty07gvU0Tbmf4Kggtr7E3_xKUOBFS2etcjUHDGloS_ql7pEiu3OwF4uCU1pdCA14pJz41EiW_gYfy5VDykMukRkYqJn2AddNmL4dKVUmeB7oo79W73TUSt7SwC05LSPCRtID-kQ7vQyCvG3MwQw4'

describe('App.vue', () => {
    let $vm

    beforeEach(() => {
        Vue.use(Vuex)
        Vue.use(localstorage)
        $vm = new Vue({
            components: {
                App,
            },
            store: Store,
            render: h => h(App),
        }).$mount()
    })

    test('should bounce to login page if no JWT exists', () => {
        expect($vm.$el).toMatchSnapshot()
        expect(typeof $vm).toBe('object')
    })

    test('should carry on loading with jwt but no cached data', async () => {
        localStorage.setItem('jwt', jwt)
        $vm.$store.commit('universal/UPDATE_JWT')

        await $vm.$store.dispatch('universal/$init').then(() => {
            expect($vm.$el).toMatchSnapshot()
        })
    })

    describe('Cached data + jwt', () => {
        test('Full temnplate', async () => {
            Vue.localForage = localforage.createInstance({ name: 'test' })
            Vue.localForage.setItem('updateUser', { id: 'test' })

            localStorage.setItem('jwt', jwt)
            $vm.$store.commit('universal/UPDATE_JWT')

            await $vm.$store.dispatch('universal/$init').then(() => {
                expect($vm.$el).toMatchSnapshot()
                Vue.localForage.removeItem('updateUser')
            })
        })

        test('empty template', async () => {
            $vm = new Vue({
                components: {
                    App,
                },
                store: Store,
                render: h => h(App, {
                    props: {
                        suppressTopbar: true,
                        suppressNav: true,
                    },
                }),
            }).$mount()

            Vue.localForage = localforage.createInstance({ name: 'test' })
            Vue.localForage.setItem('updateUser', { id: 'test' })

            localStorage.setItem('jwt', jwt)
            $vm.$store.commit('universal/UPDATE_JWT')

            await $vm.$store.dispatch('universal/$init').then(() => {
                expect($vm.$el).toMatchSnapshot()
                Vue.localForage.removeItem('updateUser')
            })
        })
    })
})
