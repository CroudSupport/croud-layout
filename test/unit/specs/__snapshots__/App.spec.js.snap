// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.vue Cached data + jwt Full temnplate 1`] = `
<div
  id="main-container"
>
  <div
    id="header"
    class="development"
  >
    <div
      id="quickJobHeader"
    >
      <div
        id="quick-selector"
      >
        <!--  -->
      </div>
    </div>
    <div
      id="loggedIn"
    >
      <a
        href="#"
      >
        <img
          class="thumbnail"
          style="vertical-align: middle;"
        />
        <strong />
      </a>
    </div>
    <div
      id="user_switcher"
    >
      <div>
        <div
          class="switcher"
        >
          <div
            class="ui floating labeled icon top right pointing dropdown tiny positive button"
            tabindex="0"
          >
            <i
              class="exchange icon"
            />
            <span
              class="text"
            >
              Switch
            </span>
            <div
              class="menu"
              tabindex="-1"
            >
              <button
                class="ui button tiny logout blue"
              >
                logout
              </button>
              <div
                class="header"
              >
                <img
                  data-id="test"
                  class="ui avatar image"
                />
                <div
                  class="label"
                />
              </div>
              <div
                class="divider"
              />
              <!--  -->
              <!--  -->
              <div
                class="scrolling menu"
              />
              <div
                class="ui inverted dimmer"
              >
                <div
                  class="ui text large loader"
                >
                  Switching User
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="wrapper"
    >
      <a
        href="/"
      >
        <h1
          id="logo"
        >
          Croud Control
        </h1>
      </a>
    </div>
  </div>
  <div
    class=""
    id="main-content"
  >
    <div
      id="main-navigation"
      class="main-navigation"
    >
      <div
        class="contained"
      >
        <!--  -->
      </div>
    </div>
    <div
      id="main-content-container"
    >
      <!--  -->
      <div
        id="main-content-body"
        class="flush-body without-heading"
      >
        <h1
          class="ui heading"
        >
          Test
        </h1>
      </div>
    </div>
  </div>
</div>
`;

exports[`App.vue Cached data + jwt empty template 1`] = `
<div
  id="main-container"
>
  <!--  -->
  <div
    class=""
    id="main-content"
  >
    <!--  -->
    <div
      id=""
    >
      <!--  -->
      <div
        id="main-content-body"
        class="flush-body without-heading"
      >
        <h1
          class="ui heading"
        >
          Test
        </h1>
      </div>
    </div>
  </div>
</div>
`;

exports[`App.vue should bounce to login page if no JWT exists 1`] = `
<div
  id="main-container"
>
  <!--  -->
  <div
    class="login-container"
  >
    <!--  -->
  </div>
</div>
`;

exports[`App.vue should carry on loading with jwt but no cached data 1`] = `
<div
  id="main-container"
>
  <!--  -->
  <div
    class="ui very padded basic center aligned loading-overlay segment"
  >
    <div>
      <img
        src="../static/img/croud_logo_new.svg"
        class="ui small image"
      />
    </div>
  </div>
</div>
`;
